In [140]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import seaborn as sns
import matplotlib.pyplot as plt
import sys
import os
sys.path.append(os.path.abspath(os.path.join('../script')))


In [141]:
from data_cleaner import Data_cleaner

In [142]:
sys.path.append(os.path.abspath(os.path.join('../script')))

In [143]:
# reading the csv file
missing_values = ["n/a", "na", "undefined"]
db = pd.read_csv('C:/Users/Hikma Brhan/User-Analytics-Telecomm-Industry/data/Week1_challenge_data_source(CSV) (1).csv', na_values=missing_values)
db.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0  9.16456699548519E+015  ...          15854611.0           2501332.0   
1                L77566A  ...          20247395.0          19111729.0   
2                D42335A  ...          19725661.0          14699576.0   
3                T21824A  ...          21388122.0          15146643.0   
4                D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 55 columns]

In [144]:
# rows and columns in the def
db.shape

(150001, 55)

In [145]:
# reading the field descriptions file
df_descript = pd.read_excel("../data/Field Descriptions.xlsx")
df_descript

Fields  \
0                                  bearer id   
1                                  Dur. (ms)   
2                                      Start   
3                                   Start ms   
4                                        End   
5                                     End ms   
6                                   Dur. (s)   
7                                       IMSI   
8                              MSISDN/Number   
9                                       IMEI   
10                        Last Location Name   
11                           Avg RTT DL (ms)   
12                           Avg RTT UL (ms)   
13                   Avg Bearer TP DL (kbps)   
14                   Avg Bearer TP UL (kbps)   
15               TCP DL Retrans. Vol (Bytes)   
16               TCP UL Retrans. Vol (Bytes)   
17                       DL TP < 50 Kbps (%)   
18            50 Kbps < DL TP < 250 Kbps (%)   
19             250 Kbps < DL TP < 1 Mbps (%)   
20                        DL TP > 1 Mbps (%)   
21                       UL TP < 10 Kbps (%)   
22             10 Kbps < UL TP < 50 Kbps (%)   
23            50 Kbps < UL TP < 300 Kbps (%)   
24                      UL TP > 300 Kbps (%)   
25                           HTTP DL (Bytes)   
26                           HTTP UL (Bytes)   
27                 Activity Duration DL (ms)   
28                 Activity Duration UL (ms)   
29                               Dur. (ms).1   
30                      Handset Manufacturer   
31                              Handset Type   
32           Nb of sec with 125000B < Vol DL   
33     Nb of sec with 1250B < Vol UL < 6250B   
34  Nb of sec with 31250B < Vol DL < 125000B   
35            Nb of sec with 37500B < Vol UL   
36    Nb of sec with 6250B < Vol DL < 31250B   
37    Nb of sec with 6250B < Vol UL < 37500B   
38             Nb of sec with Vol DL < 6250B   
39             Nb of sec with Vol UL < 1250B   
40                   Social Media DL (Bytes)   
41                   Social Media UL (Bytes)   
42                        YouTube DL (Bytes)   
43                        YouTube UL (Bytes)   
44                        Netflix DL (Bytes)   
45                        Netflix UL (Bytes)   
46                         Google DL (Bytes)   
47                         Google UL (Bytes)   
48                          Email DL (Bytes)   
49                          Email UL (Bytes)   
50                         Gaming DL (Bytes)   
51                         Gaming UL (Bytes)   
52                                  Other DL   
53                                  Other UL   
54                          Total DL (Bytes)   
55                          Total UL (Bytes)   

                                          Description  
0                              xDr session identifier  
1                   Total Duration of the xDR (in ms)  
2       Start time of the xDR (first frame timestamp)  
3   Milliseconds offset of start time for the xDR ...  
4          End time of the xDR (last frame timestamp)  
5   Milliseconds offset of end time of the xDR (la...  
6                    Total Duration of the xDR (in s)  
7            International Mobile Subscriber Identity  
8   MS International PSTN/ISDN Number of mobile - ...  
9             International Mobile Equipment Identity  
10  User location call name (2G/3G/4G) at the end ...  
11  Average Round Trip Time measurement Downlink d...  
12  Average Round Trip Time measurement Uplink dir...  
13  Average Bearer Throughput for Downlink (kbps) ...  
14  Average Bearer Throughput for uplink (kbps) - ...  
15  TCP volume of Downlink packets detected as ret...  
16  TCP volume of Uplink packets detected as retra...  
17  Duration ratio when Bearer Downlink Throughput...  
18  Duration ratio when Bearer Downlink Throughput...  
19  Duration ratio when Bearer Downlink Throughput...  
20  Duration ratio when Bearer Downlink Throughput...  
21  Duration ratio when Bearer Uplink Throughput < ….  
22  Duration 

In [146]:
db = Data_cleaner().drop_duplicate(db)
db.shape

(150001, 55)

In [147]:
db = Data_cleaner().drop_duplicate(db)
db.shape

(150001, 55)

In [153]:
db.isna().sum()

Bearer Id                                      991
Start                                            1
Start ms                                         1
End                                              1
End ms                                           1
Dur. (ms)                                        1
IMSI                                           570
MSISDN/Number                                 1066
IMEI                                           572
Last Location Name                            1153
Avg RTT DL (ms)                              27829
Avg RTT UL (ms)                              27812
Avg Bearer TP DL (kbps)                          1
Avg Bearer TP UL (kbps)                          1
TCP DL Retrans. Vol (Bytes)                  88146
TCP UL Retrans. Vol (Bytes)                  96649
DL TP < 50 Kbps (%)                            754
50 Kbps < DL TP < 250 Kbps (%)                 754
250 Kbps < DL TP < 1 Mbps (%)                  754
DL TP > 1 Mbps (%)             

In [152]:


Data_cleaner().percent_missing_values(db)

The Telecom dataset contains 12.72 % missing values.


### Missing Values

In [150]:
 Data_cleaner().percent_missing_values(db)

The Telecom dataset contains 12.72 % missing values.


In [154]:
missing_df = Data_cleaner.missing_values_table(db)

Your selected dataframe has 55 columns.
There are 41 columns that have missing values.


In [155]:
missing_df.shape

(41, 3)

In [157]:
missing_df.head(20)

Missing Values  % of Total Values  \
Nb of sec with 37500B < Vol UL                    130254              86.84   
Nb of sec with 6250B < Vol UL < 37500B            111843              74.56   
Nb of sec with 125000B < Vol DL                    97538              65.02   
TCP UL Retrans. Vol (Bytes)                        96649              64.43   
Nb of sec with 31250B < Vol DL < 125000B           93586              62.39   
Nb of sec with 1250B < Vol UL < 6250B              92894              61.93   
Nb of sec with 6250B < Vol DL < 31250B             88317              58.88   
TCP DL Retrans. Vol (Bytes)                        88146              58.76   
HTTP UL (Bytes)                                    81810              54.54   
HTTP DL (Bytes)                                    81474              54.32   
Avg RTT DL (ms)                                    27829              18.55   
Avg RTT UL (ms)                                    27812              18.54   
Handset Manufacturer                                9559               6.37   
Handset Type                                        9559               6.37   
Last Location Name                                  1153               0.77   
MSISDN/Number                                       1066               0.71   
Bearer Id                                            991               0.66   
Nb of sec with Vol UL < 1250B                        793               0.53   
10 Kbps < UL TP < 50 Kbps (%)                        792               0.53   
UL TP > 300 Kbps (%)                                 792               0.53   

                                            Dtype  
Nb of sec with 37500B < Vol UL            float64  
Nb of sec with 6250B < Vol UL < 37500B    float64  
Nb of sec with 125000B < Vol DL           float64  
TCP UL Retrans. Vol (Bytes)               float64  
Nb of sec with 31250B < Vol DL < 125000B  float64  
Nb of sec with 1250B < Vol UL < 6250B     float64  
Nb of sec with 6250B < Vol DL < 31250B    float64  
TCP DL Retrans. Vol (Bytes)               float64  
HTTP UL (Bytes)                           float64  
HTTP DL (Bytes)                           float64  
Avg RTT DL (ms)                           float64  
Avg RTT UL (ms)                           float64  
Handset Manufacturer                       object  
Handset Type                               object  
Last Location Name                         object  
MSISDN/Number                             float64  
Bearer Id                                 float64  
Nb of sec with Vol UL < 1250B             float64  
10 Kbps < UL TP < 50 Kbps (%)             float64  
UL TP > 300 Kbps (%)                      float64

#### it is good to remove columns with more than 30% of missing values.

In [158]:
missing_big = missing_df[missing_df['% of Total Values'] >=30.00].index.tolist()
missing_big

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'TCP UL Retrans. Vol (Bytes)',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'TCP DL Retrans. Vol (Bytes)',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

In [159]:
olumns_skewed = ['TCP UL Retrans. Vol (Bytes)','TCP DL Retrans. Vol (Bytes)']

In [160]:
print(db['TCP UL Retrans. Vol (Bytes)'].skew())
print(db['TCP DL Retrans. Vol (Bytes)'].skew())

84.1133930476764
15.951808879964869


In [161]:
drop_columns = [col for col in missing_big if col not in columns_skewed]
drop_columns 

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

In [162]:
db = db.drop(columns_to_remove,axis=1)

In [163]:
db.shape

(150001, 47)

In [164]:
 Data_cleaner().percent_missing_values(db)

The Telecom dataset contains 3.85 % missing values.


In [165]:
Data_cleaner.missing_values_table(db)

Your selected dataframe has 47 columns.
There are 33 columns that have missing values.


Missing Values  % of Total Values    Dtype
TCP UL Retrans. Vol (Bytes)              96649              64.43  float64
TCP DL Retrans. Vol (Bytes)              88146              58.76  float64
Avg RTT DL (ms)                          27829              18.55  float64
Avg RTT UL (ms)                          27812              18.54  float64
Handset Type                              9559               6.37   object
Handset Manufacturer                      9559               6.37   object
Last Location Name                        1153               0.77   object
MSISDN/Number                             1066               0.71  float64
Bearer Id                                  991               0.66  float64
Nb of sec with Vol UL < 1250B              793               0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792               0.53  float64
UL TP < 10 Kbps (%)                        792               0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792               0.53  float64
UL TP > 300 Kbps (%)                       792               0.53  float64
Nb of sec with Vol DL < 6250B              755               0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754               0.50  float64
DL TP > 1 Mbps (%)                         754               0.50  float64
DL TP < 50 Kbps (%)                        754               0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754               0.50  float64
IMEI                                       572               0.38  float64
IMSI                                       570               0.38  float64
Start                                        1               0.00   object
Avg Bearer TP UL (kbps)                      1               0.00  float64
Avg Bearer TP DL (kbps)                      1               0.00  float64
Activity Duration DL (ms)                    1               0.00  float64
Activity Duration UL (ms)                    1               0.00  float64
Dur. (ms).1                                  1               0.00  float64
Dur. (ms)                                    1               0.00  float64
End ms                                       1               0.00  float64
End                                          1               0.00   object
Start ms                                     1               0.00  float64
Total UL (Bytes)                             1               0.00  float64
Total DL (Bytes)                             1               0.00  float64

In [168]:
Data_cleaner.fix_missing_bfill(db,['TCP UL Retrans. Vol (Bytes)','TCP DL Retrans. Vol (Bytes)'])

0         19520.0
1         19520.0
2         19520.0
3         19520.0
4         19520.0
           ...   
149996        NaN
149997        NaN
149998        NaN
149999        NaN
150000        NaN
Name: TCP DL Retrans. Vol (Bytes), Length: 150001, dtype: float64

### Avg RTT DL (ms) and Avg RTT UL (ms) have 18.54% and 18.55% respectively

In [169]:
print(db['Avg RTT DL (ms)'].skew(skipna=True))
print(db['Avg RTT UL (ms)'].skew(skipna=True))

62.90782807995961
28.45741458546382


In [170]:
db= Data_cleaner.fix_missing_ffill(db,['Avg RTT DL (ms)','Avg RTT UL (ms)'])

In [171]:
Data_cleaner().percent_missing_values(db)

The Telecom dataset contains 0.44 % missing values.


In [172]:
Data_cleaner.missing_values_table(db)

Your selected dataframe has 47 columns.
There are 31 columns that have missing values.


Missing Values  % of Total Values    Dtype
Handset Type                              9559               6.37   object
Handset Manufacturer                      9559               6.37   object
Last Location Name                        1153               0.77   object
MSISDN/Number                             1066               0.71  float64
Bearer Id                                  991               0.66  float64
Nb of sec with Vol UL < 1250B              793               0.53  float64
UL TP > 300 Kbps (%)                       792               0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792               0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792               0.53  float64
UL TP < 10 Kbps (%)                        792               0.53  float64
Nb of sec with Vol DL < 6250B              755               0.50  float64
DL TP < 50 Kbps (%)                        754               0.50  float64
DL TP > 1 Mbps (%)                         754               0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754               0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754               0.50  float64
IMEI                                       572               0.38  float64
IMSI                                       570               0.38  float64
TCP UL Retrans. Vol (Bytes)                  5               0.00  float64
TCP DL Retrans. Vol (Bytes)                  5               0.00  float64
Start                                        1               0.00   object
Avg Bearer TP UL (kbps)                      1               0.00  float64
Avg Bearer TP DL (kbps)                      1               0.00  float64
Activity Duration DL (ms)                    1               0.00  float64
Activity Duration UL (ms)                    1               0.00  float64
Dur. (ms).1                                  1               0.00  float64
Dur. (ms)                                    1               0.00  float64
End ms                                       1               0.00  float64
End                                          1               0.00   object
Start ms                                     1               0.00  float64
Total UL (Bytes)                             1               0.00  float64
Total DL (Bytes)                             1               0.00  float64

### Fill the missing values of Handset Type and Handset Manufacturer

#### Handset Type and handset Manufacturer are categorical columns , it is better to fill them with "unknown".

In [173]:
Data_cleaner.fix_missing_value(db, 'Handset Type', 'unknown')
Data_cleaner.fix_missing_value(db, 'Handset Manufacturer', 'unknown')

9559 missing values in the column Handset Type have been replaced by unknown.
9559 missing values in the column Handset Manufacturer have been replaced by unknown.


0         Samsung
1         Samsung
2         Samsung
3         unknown
4         Samsung
           ...   
149996      Apple
149997      Apple
149998      Apple
149999     Huawei
150000    unknown
Name: Handset Manufacturer, Length: 150001, dtype: object

In [174]:
Data_cleaner().percent_missing_values(db)

The Telecom dataset contains 0.17 % missing values.


#### we are left with small persentage of null values. so it is better if we remove the from data.

In [175]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 150001 entries, 0 to 150000
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Bearer Id                       149010 non-null  float64
 1   Start                           150000 non-null  object 
 2   Start ms                        150000 non-null  float64
 3   End                             150000 non-null  object 
 4   End ms                          150000 non-null  float64
 5   Dur. (ms)                       150000 non-null  float64
 6   IMSI                            149431 non-null  float64
 7   MSISDN/Number                   148935 non-null  float64
 8   IMEI                            149429 non-null  float64
 9   Last Location Name              148848 non-null  object 
 10  Avg RTT DL (ms)                 150001 non-null  float64
 11  Avg RTT UL (ms)                 150001 non-null  float64
 12  Avg Bearer TP DL

In [176]:
missing_big = missing_df[missing_df['% of Total Values'] >=30.00].index.tolist()
missing_big

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'TCP UL Retrans. Vol (Bytes)',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'TCP DL Retrans. Vol (Bytes)',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

In [177]:
db.dropna(inplace=True)

In [178]:
Data_cleaner().percent_missing_values(db)

The Telecom dataset contains 0.0 % missing values.


### Data Types

In [179]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146887 entries, 0 to 149995
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Bearer Id                       146887 non-null  float64
 1   Start                           146887 non-null  object 
 2   Start ms                        146887 non-null  float64
 3   End                             146887 non-null  object 
 4   End ms                          146887 non-null  float64
 5   Dur. (ms)                       146887 non-null  float64
 6   IMSI                            146887 non-null  float64
 7   MSISDN/Number                   146887 non-null  float64
 8   IMEI                            146887 non-null  float64
 9   Last Location Name              146887 non-null  object 
 10  Avg RTT DL (ms)                 146887 non-null  float64
 11  Avg RTT UL (ms)                 146887 non-null  float64
 12  Avg Bearer TP DL

In [181]:
db

Bearer Id            Start  Start ms              End  End ms  \
0       1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1       1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2       1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3       1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4       1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
...              ...              ...       ...              ...     ...   
149991  7.349883e+18   4/29/2019 7:28     794.0   4/30/2019 0:36   523.0   
149992  1.311448e+19   4/29/2019 7:28     114.0   4/30/2019 7:04   724.0   
149993  1.311448e+19   4/29/2019 7:28      79.0  4/30/2019 18:22   512.0   
149994  1.311448e+19   4/29/2019 7:28      83.0  4/30/2019 15:01   268.0   
149995  1.304243e+19   4/29/2019 7:28     615.0   4/30/2019 0:01   407.0   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1       1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2       1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3       1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4       1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   
...           ...           ...            ...           ...   
149991    61661.0  2.082017e+14   3.376215e+10  3.586061e+13   
149992    84940.0  2.082014e+14   3.376127e+10  3.532701e+13   
149993   125622.0  2.082014e+14   3.362611e+10  3.573531e+13   
149994   113545.0  2.082003e+14   3.364566e+10  3.515541e+13   
149995    59587.0  2.082014e+14   3.366865e+10  3.533251e+13   

           Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0       9.16456699548519E+015  ...          15854611.0           2501332.0   
1                     L77566A  ...          20247395.0          19111729.0   
2                     D42335A  ...          19725661.0          14699576.0   
3                     T21824A  ...          21388122.0          15146643.0   
4                     D88865A  ...          15259380.0          18962873.0   
...                       ...  ...                 ...                 ...   
149991                D10033B  ...           7404819.0           9864285.0   
149992                D78058B  ...          22428728.0          14385815.0   
149993                D73542B  ...          18144703.0           8161828.0   
149994                T88383B  ...          23043782.0          19176074.0   
149995                T85721A  ...           6550499.0          18003146.0   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0                8198936.0           9656251.0        278082303.0   
1               18338413.0          17227132.0        608750074.0   
2               17587794.0           6163408.0        229584621.0   
3               13994646.0           1097942.0        799538153.0   
4               17124581.0            415218.0        527707248.0   
...                    ...                 ...                ...   
149991          18954727.0          19382787.0        258688968.0   
149992           6940672.0           6801943.0         35738570.0   
149993          20559422.0          20415237.0        569668054.0   
149994          18762809.0           9591310.0        810666072.0   
149995          22468983.0           7149728.0        784435351.0   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0              14344150.0       171744450.0         8814393.0   
1               1170709.0       526904238.0        15055145.0   
2                395630.0       410692588.0         4215763.0   
3              10849722.0       749039933.0        12797283.0   
4               3529801.0       550709500.0        13910322.0   
...                   ...               ...               ...   
149991          9542486.0       249358264.0        14134902.0   
149992          7842728.0       808482329.0

### Data type for Start and End Columns of the dataset is so lets as convet it to date time datatype

In [203]:
Data_cleaner.convert_to_datetime(db,['Start','End'])

Bearer Id               Start  Start ms                 End  \
0      -9223372036854775808 2019-04-04 12:01:00     770.0 2019-04-25 14:35:00   
1      -9223372036854775808 2019-04-09 13:04:00     235.0 2019-04-25 08:15:00   
2      -9223372036854775808 2019-04-09 17:42:00       1.0 2019-04-25 11:58:00   
3      -9223372036854775808 2019-04-10 00:31:00     486.0 2019-04-25 07:36:00   
4      -9223372036854775808 2019-04-12 20:10:00     565.0 2019-04-25 10:40:00   
...                     ...                 ...       ...                 ...   
149991  7349883264234529792 2019-04-29 07:28:00     794.0 2019-04-30 00:36:00   
149992 -9223372036854775808 2019-04-29 07:28:00     114.0 2019-04-30 07:04:00   
149993 -9223372036854775808 2019-04-29 07:28:00      79.0 2019-04-30 18:22:00   
149994 -9223372036854775808 2019-04-29 07:28:00      83.0 2019-04-30 15:01:00   
149995 -9223372036854775808 2019-04-29 07:28:00     615.0 2019-04-30 00:01:00   

        End ms  Dur. (ms)             IMSI  MSISDN/Number            IMEI  \
0        662.0  1823652.0  208201448079117    33664962239  35521209507511   
1        606.0  1365104.0  208201909211140    33681854413  35794009006359   
2        652.0  1361762.0  208200314458056    33760627129  35281510359387   
3        171.0  1321509.0  208201402342131    33750343200  35356610164913   
4        954.0  1089009.0  208201401415120    33699795932  35407009745539   
...        ...        ...              ...            ...             ...   
149991   523.0    61661.0  208201716588303    33762146363  35860607430880   
149992   724.0    84940.0  208201402149148    33761274518  35327007109396   
149993   512.0   125622.0  208201402072093    33626114091  35735309473406   
149994   268.0   113545.0  208200313074099    33645655643  35155409288696   
149995   407.0    59587.0  208201402604057    33668648496  35332510508318   

           Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0       9.16456699548519E+015  ...          15854611.0           2501332.0   
1                     L77566A  ...          20247395.0          19111729.0   
2                     D42335A  ...          19725661.0          14699576.0   
3                     T21824A  ...          21388122.0          15146643.0   
4                     D88865A  ...          15259380.0          18962873.0   
...                       ...  ...                 ...                 ...   
149991                D10033B  ...           7404819.0           9864285.0   
149992                D78058B  ...          22428728.0          14385815.0   
149993                D73542B  ...          18144703.0           8161828.0   
149994                T88383B  ...          23043782.0          19176074.0   
149995                T85721A  ...           6550499.0          18003146.0   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0                8198936.0           9656251.0        278082303.0   
1               18338413.0          17227132.0        608750074.0   
2               17587794.0           6163408.0        229584621.0   
3               13994646.0           1097942.0        799538153.0   
4               17124581.0            415218.0        527707248.0   
...                    ...                 ...                ...   
149991          18954727.0          19382787.0        258688968.0   
149992           6940672.0           6801943.0         35738570.0   
149993          20559422.0          20415237.0        569668054.0   
149994          18762809.0           9591310.0        810666072.0   
149995          22468983.0           7149728.0        784435351.0   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0              14344150.0       171744450.0         8814393.0   
1               1170709.0       526904238.0        15055145.0   
2                395630.0       410692588.0         4215763.0   
3              10849722.0       749039933.0        12797283.0   
4               3529801.0    

In [204]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146887 entries, 0 to 149995
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Bearer Id                       146887 non-null  int64         
 1   Start                           146887 non-null  datetime64[ns]
 2   Start ms                        146887 non-null  float64       
 3   End                             146887 non-null  datetime64[ns]
 4   End ms                          146887 non-null  float64       
 5   Dur. (ms)                       146887 non-null  float64       
 6   IMSI                            146887 non-null  int64         
 7   MSISDN/Number                   146887 non-null  int64         
 8   IMEI                            146887 non-null  int64         
 9   Last Location Name              146887 non-null  string        
 10  Avg RTT DL (ms)                 146887 non-null  float64

### Now change the column with object datatype to stirng

In [205]:
# get the columns with object data type
string_columns = db.select_dtypes(include='object').columns.tolist()
string_columns

[]

In [206]:
Data_cleaner.convert_to_string(db,string_columns)

In [207]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146887 entries, 0 to 149995
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Bearer Id                       146887 non-null  int64         
 1   Start                           146887 non-null  datetime64[ns]
 2   Start ms                        146887 non-null  float64       
 3   End                             146887 non-null  datetime64[ns]
 4   End ms                          146887 non-null  float64       
 5   Dur. (ms)                       146887 non-null  float64       
 6   IMSI                            146887 non-null  int64         
 7   MSISDN/Number                   146887 non-null  int64         
 8   IMEI                            146887 non-null  int64         
 9   Last Location Name              146887 non-null  string        
 10  Avg RTT DL (ms)                 146887 non-null  float64

### Convet for the identification numbert to int

In [212]:
#Data_cleaner.convert_to_int(db, ["Bearer Id","IMSI","MSISDN/Number", "IMEI" ])

In [213]:
db

Bearer Id               Start  Start ms                 End  \
0      -9223372036854775808 2019-04-04 12:01:00     770.0 2019-04-25 14:35:00   
1      -9223372036854775808 2019-04-09 13:04:00     235.0 2019-04-25 08:15:00   
2      -9223372036854775808 2019-04-09 17:42:00       1.0 2019-04-25 11:58:00   
3      -9223372036854775808 2019-04-10 00:31:00     486.0 2019-04-25 07:36:00   
4      -9223372036854775808 2019-04-12 20:10:00     565.0 2019-04-25 10:40:00   
...                     ...                 ...       ...                 ...   
149991  7349883264234529792 2019-04-29 07:28:00     794.0 2019-04-30 00:36:00   
149992 -9223372036854775808 2019-04-29 07:28:00     114.0 2019-04-30 07:04:00   
149993 -9223372036854775808 2019-04-29 07:28:00      79.0 2019-04-30 18:22:00   
149994 -9223372036854775808 2019-04-29 07:28:00      83.0 2019-04-30 15:01:00   
149995 -9223372036854775808 2019-04-29 07:28:00     615.0 2019-04-30 00:01:00   

        End ms  Dur. (ms)             IMSI  MSISDN/Number            IMEI  \
0        662.0  1823652.0  208201448079117    33664962239  35521209507511   
1        606.0  1365104.0  208201909211140    33681854413  35794009006359   
2        652.0  1361762.0  208200314458056    33760627129  35281510359387   
3        171.0  1321509.0  208201402342131    33750343200  35356610164913   
4        954.0  1089009.0  208201401415120    33699795932  35407009745539   
...        ...        ...              ...            ...             ...   
149991   523.0    61661.0  208201716588303    33762146363  35860607430880   
149992   724.0    84940.0  208201402149148    33761274518  35327007109396   
149993   512.0   125622.0  208201402072093    33626114091  35735309473406   
149994   268.0   113545.0  208200313074099    33645655643  35155409288696   
149995   407.0    59587.0  208201402604057    33668648496  35332510508318   

           Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0       9.16456699548519E+015  ...          15854611.0           2501332.0   
1                     L77566A  ...          20247395.0          19111729.0   
2                     D42335A  ...          19725661.0          14699576.0   
3                     T21824A  ...          21388122.0          15146643.0   
4                     D88865A  ...          15259380.0          18962873.0   
...                       ...  ...                 ...                 ...   
149991                D10033B  ...           7404819.0           9864285.0   
149992                D78058B  ...          22428728.0          14385815.0   
149993                D73542B  ...          18144703.0           8161828.0   
149994                T88383B  ...          23043782.0          19176074.0   
149995                T85721A  ...           6550499.0          18003146.0   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0                8198936.0           9656251.0        278082303.0   
1               18338413.0          17227132.0        608750074.0   
2               17587794.0           6163408.0        229584621.0   
3               13994646.0           1097942.0        799538153.0   
4               17124581.0            415218.0        527707248.0   
...                    ...                 ...                ...   
149991          18954727.0          19382787.0        258688968.0   
149992           6940672.0           6801943.0         35738570.0   
149993          20559422.0          20415237.0        569668054.0   
149994          18762809.0           9591310.0        810666072.0   
149995          22468983.0           7149728.0        784435351.0   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0              14344150.0       171744450.0         8814393.0   
1               1170709.0       526904238.0        15055145.0   
2                395630.0       410692588.0         4215763.0   
3              10849722.0       749039933.0        12797283.0   
4               3529801.0    

In [214]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146887 entries, 0 to 149995
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Bearer Id                       146887 non-null  int64         
 1   Start                           146887 non-null  datetime64[ns]
 2   Start ms                        146887 non-null  float64       
 3   End                             146887 non-null  datetime64[ns]
 4   End ms                          146887 non-null  float64       
 5   Dur. (ms)                       146887 non-null  float64       
 6   IMSI                            146887 non-null  int64         
 7   MSISDN/Number                   146887 non-null  int64         
 8   IMEI                            146887 non-null  int64         
 9   Last Location Name              146887 non-null  string        
 10  Avg RTT DL (ms)                 146887 non-null  float64

In [215]:
db.to_csv('../data/preprocessed_data.csv')

In [216]:
db.head()

Bearer Id               Start  Start ms                 End  \
0 -9223372036854775808 2019-04-04 12:01:00     770.0 2019-04-25 14:35:00   
1 -9223372036854775808 2019-04-09 13:04:00     235.0 2019-04-25 08:15:00   
2 -9223372036854775808 2019-04-09 17:42:00       1.0 2019-04-25 11:58:00   
3 -9223372036854775808 2019-04-10 00:31:00     486.0 2019-04-25 07:36:00   
4 -9223372036854775808 2019-04-12 20:10:00     565.0 2019-04-25 10:40:00   

   End ms  Dur. (ms)             IMSI  MSISDN/Number            IMEI  \
0   662.0  1823652.0  208201448079117    33664962239  35521209507511   
1   606.0  1365104.0  208201909211140    33681854413  35794009006359   
2   652.0  1361762.0  208200314458056    33760627129  35281510359387   
3   171.0  1321509.0  208201402342131    33750343200  35356610164913   
4   954.0  1089009.0  208201401415120    33699795932  35407009745539   

      Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0  9.16456699548519E+015  ...          15854611.0           2501332.0   
1                L77566A  ...          20247395.0          19111729.0   
2                D42335A  ...          19725661.0          14699576.0   
3                T21824A  ...          21388122.0          15146643.0   
4                D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 47 columns]